### PR TITLE
Switch to constraint based collision solving

### PR DESCRIPTION
**Pull request by fallenatlas**
Replace basic collision solving to proper XPBD constraint based collision solving.

TIP: For anyone who may try to run the new sample, compile it on release... otherwise your pc might blow up
(or change the number of cubes in the sample)

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: engine/src/physics/solver/penetration_constraint/plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#include "../../../collisions/narrow_phase/plugin.hpp"
```
Comment (RiscadoA) [inline engine/src/physics/solver/penetration_constraint/plugin.cpp:1]: After rebasing on main you can use the `collisionsTag`, which I added due to previously for the shared lib PR, so that we didn't need to access the private headers. Thus, you can remove this include then.

TL;DR physics shouldn't care about collisions internals

File: engine/src/physics/solver/penetration_constraint/plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                                                  .entity1OriginalPosition = collidingWith.entity1OriginalPosition,
+                                                  .entity2OriginalPosition = collidingWith.entity2OriginalPosition,
+                                                  .penetration = collidingWith.penetration,
+                                                  .normal = collidingWith.normal});
+            }
+        });
+
+    cubos.system("clean penetration constraint pairs")
+        .tagged(penetrationConstraintCleanTag)
+        .before(collisionsNarrowTag)
```
Comment (RiscadoA) [inline engine/src/physics/solver/penetration_constraint/plugin.cpp:119]: Shouldn't these two systems be in fixedStep / physicsSolve?

File: engine/src/physics/solver/penetration_constraint/plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                }
+
+                constraint.lagrangeMultiplier += lagrangeUpdate;
+            }
+        });
+
+    cubos.system("clear penetration constraint lagrange multiplier")
+        .tagged(physicsSimulationClearForcesTag)
+        .after(physicsSimulationSubstepsUpdateVelocityTag)
+        .tagged(fixedStepTag)
```
Comment (RiscadoA) [inline engine/src/physics/solver/penetration_constraint/plugin.cpp:93]: no physicsSolve tag?

File: engine/src/physics/solver/penetration_constraint/plugin.hpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/// @file
+/// @brief Penetration Constraint plugin, resource @ref cubos::engine::PenetrationConstraintPlugin.
```
Comment (RiscadoA) [inline engine/src/physics/solver/penetration_constraint/plugin.hpp:2]:
```suggestion
/// @brief Penetration Constraint plugin.
```

File: engine/src/physics/solver/plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#include "../../collisions/narrow_phase/plugin.hpp"
```
Comment (RiscadoA) [inline engine/src/physics/solver/plugin.cpp:1]: Unnecessary

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! :partying_face: I will trust your math, just requesting some changes on programming stuff

File: engine/src/physics/fixed_substep/substeps.cpp [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#include "substeps.hpp"
+
+#include <cubos/core/ecs/reflection.hpp>
+
+CUBOS_REFLECT_IMPL(cubos::engine::Substeps)
+{
+    return core::ecs::TypeBuilder<Substeps>("cubos::engine::Substeps").build();
```
Comment (RiscadoA) [inline engine/src/physics/fixed_substep/substeps.cpp:7]:
```suggestion
#include <cubos/core/ecs/reflection.hpp>
#include <cubos/core/reflection/external/primitives.hpp>

CUBOS_REFLECT_IMPL(cubos::engine::Substeps)
{
    return core::ecs::TypeBuilder<Substeps>("cubos::engine::Substeps").withField("value", &Substeps::value).build();
```

**Review by RiscadoA — APPROVED** [review]
LGTM! Just one minor thing to fix, and missing a change log entry. You can merge it afterwards. Thanks for working on this :partying_face:
